private/pkg/storage/storagemanifest/bucket.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storagemanifest
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/bufbuild/buf/private/pkg/manifest"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"github.com/bufbuild/buf/private/pkg/storage/storageutil"
+)
+
+type bucket struct {
+	manifest *manifest.Manifest
+	blobSet  *manifest.BlobSet
+}
+
+func newBucket(
+	m *manifest.Manifest,
+	blobSet *manifest.BlobSet,
+	options ...ReadBucketOption,
+) (*bucket, error) {
+	readBucketOptions := newReadBucketOptions()
+	for _, option := range options {
+		option(readBucketOptions)
+	}
+	// TODO: why is this validation in newBucket, why is this not somewhere else?
+	// This should not be in storagemanifest.
```

Comment:
This was to allow partial manifests buckets, so we could build a bucket with just some of the blobs instead of all of them. I don't recall if we're actually using it somewhere, I can check. I guess we could move these validations where needed, somewhere else before building the bucket, but I don't see the harm, it's a different builder with its custom options, why not offering them as new bucket options?

Reply:
This validation has nothing to do with bucket construction - it has to do with the `<Manifest, BlobSet>` pair, which, if they are so tightly correlated in such a way, should probably have a pair type that does such validation.